cryptography/src/bls12381/primitives/ops.rs

```diff
@@ -496,301 +490,345 @@ where
             messages
                 .into_par_iter()
                 .map(|(namespace, msg)| {
-                    let mut hm = group::Signature::zero();
+                    let mut hm = V::Signature::zero();
                     match namespace {
-                        Some(namespace) => hm.map(MESSAGE, &union_unique(namespace, msg)),
-                        None => hm.map(MESSAGE, msg),
+                        Some(namespace) => hm.map(V::MESSAGE, &union_unique(namespace, msg)),
+                        None => hm.map(V::MESSAGE, msg),
                     };
                     hm
                 })
-                .reduce(group::Signature::zero, |mut sum, hm| {
+                .reduce(V::Signature::zero, |mut sum, hm| {
                     sum.add(&hm);
                     sum
                 })
         })
     };
 
     // Verify the signature
-    if !equal(public, signature, &hm_sum) {
-        return Err(Error::InvalidSignature);
-    }
-    Ok(())
+    V::verify_prehashed(public, &hm_sum, signature)
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::bls12381::dkg::ops::generate_shares;
+    use crate::bls12381::{
+        dkg::ops::generate_shares,
+        primitives::variant::{MinPk, MinSig},
+    };
     use blst::BLST_ERROR;
     use commonware_codec::DecodeExt;
     use commonware_utils::quorum;
-    use group::{G1, G1_MESSAGE, G1_PROOF_OF_POSSESSION, G2};
+    use group::{Private, G1_MESSAGE, G2_MESSAGE};
     use poly::Poly;
     use rand::prelude::*;
 
-    #[test]
-    fn test_codec() {
+    fn codec<V: Variant>() {
         // Encode private/public key
-        let (private, public) = keypair(&mut thread_rng());
+        let (private, public) = keypair::<_, V>(&mut thread_rng());
         let (private_bytes, public_bytes) = (private.encode(), public.encode());
 
         // Decode private/public key
         let (private_decoded, public_decoded) = (
             group::Private::decode(private_bytes.clone()).unwrap(),
-            group::Public::decode(public_bytes.clone()).unwrap(),
+            V::Public::decode(public_bytes.clone()).unwrap(),
         );
 
         // Ensure equal
         assert_eq!(private, private_decoded);
         assert_eq!(public, public_decoded);
 
         // Ensure blst compatibility
-        blst::min_pk::SecretKey::from_bytes(&private_bytes).unwrap();
-        let blst_public_decoded = blst::min_pk::PublicKey::from_bytes(&public_bytes).unwrap();
-        blst_public_decoded.validate().unwrap();
-        let blst_public_encoded = blst_public_decoded.compress().to_vec();
-        assert_eq!(public_bytes, blst_public_encoded.as_slice());
+        match V::MESSAGE {
+            G1_MESSAGE => {
+                blst::min_sig::SecretKey::from_bytes(&private_bytes).unwrap();
```

Comment:
We don't use the `::min_sig::` pattern (like `blst`) because it doesn't allow for us to build mechanisms that can be generically used.